apps/service_providers/llm_service/helper.py

```diff
@@ -1,43 +1,92 @@
+import json
 import uuid
+from json import JSONDecodeError
 
 from langchain.agents.output_parsers.tools import ToolAgentAction, parse_ai_message_to_tool_action
-from langchain_core.agents import AgentFinish
+from langchain_core.agents import AgentAction, AgentFinish
+from langchain_core.exceptions import OutputParserException
+from langchain_core.messages import AIMessage
 
 original_parse = parse_ai_message_to_tool_action
 
 
-def custom_parse_ai_message(message):
-    try:
-        return original_parse(message)
-
-    except Exception as e:
-        if "tool_call_id" in str(e) and "NoneType" in str(e):
-            tool_calls = message.additional_kwargs.get("tool_calls", [])
-
-            if tool_calls:
-                fixed_actions = []
-                for call in tool_calls:
-                    function = call.get("function", {})
-                    function_name = function.get("name", "")
-                    try:
-                        import json
-
-                        args = json.loads(function.get("arguments", "{}"))
-                    except json.JSONDecodeError:
-                        args = {}
-
-                    fixed_actions.append(
-                        ToolAgentAction(
-                            tool=function_name,
-                            tool_input=args,
-                            log=f"Invoking: {function_name} with {args}",
-                            message_log=[message],
-                            tool_call_id=f"claude_tool_{uuid.uuid4()}",
-                        )
-                    )
-                return fixed_actions
-
-            # No tool calls treat it as final response
-            return AgentFinish(return_values={"output": message.content}, log=str(message.content))
-        else:
-            raise
+def custom_parse_ai_message(message) -> list[AgentAction] | AgentFinish:
+    """Parse an AI message potentially containing tool_calls.
+    This function is adapted from the `parse_ai_message_to_tool_action`
+    function in the langchain library.
+    1. Added a specific condition to skip malformed built in tool calls that
+       originate from Anthropic models (where `call.get("type") == "tool_call"`,
+       `call.get("name") == ""`, and `call.get("id") is None`).
+    2. Ensure that tool_call_id is present in every tool_call as tool_call_id is not returned by anthropic
+    built_in_tools
+    3. Separate processing of custom tools and built in tools as custom tools are present in tool_calls
+    whereas built_in_tools are present in additional kwargs
```

Comment:
These 3 cases should be tested directly. The current test appears to only test point 3